File: sdk-python/gatedhouse/events/handler.py
"""Event handler — processes Citadel and Sphinx events to keep local caches in sync."""

from __future__ import annotations

import logging
from datetime import datetime

from gatedhouse.core.config import ResolvedConfig
from gatedhouse.core.membership.cache import MembershipCache
from gatedhouse.core.delegation.cache import DelegationCache
from gatedhouse.core.roles.repository import RoleRepository
from gatedhouse.core.roles.assignment import RoleAssignmentManager
from gatedhouse.core.roles.resolver import PermissionResolver
from gatedhouse.core.types import CachedDelegation, CachedMembership, GatehouseEvent
from gatedhouse.events import types as ET

logger = logging.getLogger("gatedhouse.events.handler")


class EventHandlerRegistry:
    """Processes incoming events to keep local caches in sync."""

    def __init__(
        self,
        membership_cache: MembershipCache,
        delegation_cache: DelegationCache,
        role_repo: RoleRepository,
        role_assignments: RoleAssignmentManager,
        permission_resolver: PermissionResolver,
        config: ResolvedConfig,
    ) -> None:
        self._membership_cache = membership_cache
        self._delegation_cache = delegation_cache
        self._role_repo = role_repo
        self._role_assignments = role_assignments
        self._permission_resolver = permission_resolver
        self._config = config

    async def handle(self, event: GatehouseEvent) -> None:
        """Process an incoming event. Idempotent by design."""
        handlers = {
            ET.ORG_CREATED: self._handle_org_created,
            ET.ORG_DELETED: self._handle_org_deleted,
            ET.ORG_SUSPENDED: self._handle_org_suspended,
            ET.ORG_REACTIVATED: self._handle_org_reactivated,
            ET.MEMBERSHIP_CREATED: self._handle_membership_created,
            ET.MEMBERSHIP_UPDATED: self._handle_membership_updated,
            ET.MEMBERSHIP_SUSPENDED: self._handle_membership_suspended,
            ET.MEMBERSHIP_REACTIVATED: self._handle_membership_reactivated,
            ET.MEMBERSHIP_REMOVED: self._handle_membership_removed,
            ET.GROUP_MEMBER_ADDED: self._handle_group_member_added,
            ET.GROUP_MEMBER_REMOVED: self._handle_group_member_removed,
            ET.GROUP_DELETED: self._handle_group_deleted,
            ET.DELEGATION_CREATED: self._handle_delegation_created,
            ET.DELEGATION_REVOKED: self._handle_delegation_revoked,
            ET.DELEGATION_EXPIRED: self._handle_delegation_expired,
            ET.DELEGATION_EXHAUSTED: self._handle_delegation_exhausted,
            ET.AGENT_DEACTIVATED: self._handle_agent_deactivated,
        }

        handler = handlers.get(event.type)
        if handler:
            await handler(event)
        else:
            logger.debug("Unknown event type %s, ignoring", event.type)

    # ─── Organization Handlers ──────────────────────────────────────

    async def _handle_org_created(self, event: GatehouseEvent) -> None:
        org_id = event.data["org_id"]
        logger.info("Handling org.created: %s", org_id)
        await self._role_repo.seed_base_roles(org_id, self._config.base_roles)

    async def _handle_org_deleted(self, event: GatehouseEvent) -> None:
        org_id = event.data["org_id"]
        logger.info("Handling org.deleted: %s — purging all data", org_id)
        await self._membership_cache.remove_all_for_org(org_id)
        await self._role_assignments.delete_all_for_org(org_id)
        await self._role_assignments.delete_all_group_roles_for_org(org_id)
        await self._role_repo.delete_all_for_org(org_id)
        await self._delegation_cache.remove_all_for_org(org_id)

    async def _handle_org_suspended(self, event: GatehouseEvent) -> None:
        org_id = event.data["org_id"]
        logger.info("Handling org.suspended: %s", org_id)
        await self._membership_cache.suspend_all_for_org(org_id)

    async def _handle_org_reactivated(self, event: GatehouseEvent) -> None:
        org_id = event.data["org_id"]
        logger.info("Handling org.reactivated: %s", org_id)
        await self._membership_cache.reactivate_all_for_org(org_id)

    # ─── Membership Handlers ───────────────────────────────────────

    async def _handle_membership_created(self, event: GatehouseEvent) -> None:
        data = event.data
        membership_id = data["membership_id"]
        org_id = data["org_id"]
        groups = data.get("groups", [])

        logger.info("Handling membership.created: %s in %s", membership_id, org_id)

        await self._membership_cache.upsert(CachedMembership(
            membership_id=membership_id,
            org_id=org_id,
            entity_type=data["entity_type"],
            entity_id=data["entity_id"],
            is_owner=data.get("is_owner", False),
            status=data.get("status", "active"),
            groups=groups,
        ))

        if self._config.default_role:
            await self._role_assignments.assign(
                membership_id, self._config.default_role, org_id
            )

        if data.get("is_owner"):
            await self._role_assignments.assign(membership_id, "owner", org_id)

        await self._permission_resolver.rebuild_for_membership(
            membership_id, org_id, groups
        )

    async def _handle_membership_updated(self, event: GatehouseEvent) -> None:
        data = event.data
        membership_id = data["membership_id"]
        org_id = data["org_id"]
        groups = data.get("groups", [])

        logger.info("Handling membership.updated: %s", membership_id)

        await self._membership_cache.upsert(CachedMembership(
            membership_id=membership_id,
            org_id=org_id,
            entity_type=data["entity_type"],
            entity_id=data["entity_id"],
            is_owner=data.get("is_owner", False),
            status=data.get("status", "active"),
            groups=groups,
        ))

        await self._permission_resolver.rebuild_for_membership(
            membership_id, org_id, groups
        )

    async def _handle_membership_suspended(self, event: GatehouseEvent) -> None:
        membership_id = event.data["membership_id"]
        logger.info("Handling membership.suspended: %s", membership_id)
        await self._membership_cache.update_status(membership_id, "suspended")

    async def _handle_membership_reactivated(self, event: GatehouseEvent) -> None:
        membership_id = event.data["membership_id"]
        logger.info("Handling membership.reactivated: %s", membership_id)
        await self._membership_cache.update_status(membership_id, "active")

    async def _handle_membership_removed(self, event: GatehouseEvent) -> None:
        membership_id = event.data["membership_id"]
        logger.info("Handling membership.removed: %s", membership_id)
        await self._membership_cache.remove(membership_id)
        await self._role_assignments.delete_all_for_membership(membership_id)
        await self._permission_resolver.clear_for_membership(membership_id)

    # ─── Group Handlers ────────────────────────────────────────────

    async def _handle_group_member_added(self, event: GatehouseEvent) -> None:
        membership_id = event.data["membership_id"]
        group_id = event.data["group_id"]
        org_id = event.data["org_id"]

        logger.info("Handling group.member.added: %s -> %s", membership_id, group_id)

        await self._membership_cache.add_group(membership_id, group_id)
        cached = await self._membership_cache.find_by_id(membership_id)
        if cached:
            await self._permission_resolver.rebuild_for_membership(
                membership_id, org_id, cached.groups
            )

    async def _handle_group_member_removed(self, event: GatehouseEvent) -> None:
        membership_id = event.data["membership_id"]
        group_id = event.data["group_id"]
        org_id = event.data["org_id"]

        logger.info("Handling group.member.removed: %s from %s", membership_id, group_id)

        await self._membership_cache.remove_group(membership_id, group_id)
        cached = await self._membership_cache.find_by_id(membership_id)
        if cached:
            await self._permission_resolver.rebuild_for_membership(
                membership_id, org_id, cached.groups
            )

    async def _handle_group_deleted(self, event: GatehouseEvent) -> None:
        group_id = event.data["group_id"]
        org_id = event.data["org_id"]

        logger.info("Handling group.deleted: %s", group_id)

        await self._membership_cache.remove_group_from_all(group_id)
        await self._role_assignments.delete_all_for_group(group_id)

        memberships = await self._membership_cache.list_by_org(org_id)
        for m in memberships:
            await self._permission_resolver.rebuild_for_membership(
                m.membership_id, org_id, m.groups
            )

    # ─── Delegation Handlers ───────────────────────────────────────

    async def _handle_delegation_created(self, event: GatehouseEvent) -> None:
        data = event.data
        logger.info("Handling delegation.created: %s", data["delegation_id"])

        await self._delegation_cache.upsert(CachedDelegation(
            delegation_id=data["delegation_id"],
            agent_id=data["agent_id"],
            delegator_id=data["delegator_id"],
            delegator_membership_id=data["delegator_membership_id"],
            org_id=data["org_id"],
            scopes=data.get("scopes", []),
            constraints=data.get("constraints", {}),
            max_uses=data.get("max_uses"),
            use_count=0,
            status="active",
            expires_at=datetime.fromisoformat(data["expires_at"]),
        ))

    async def _handle_delegation_revoked(self, event: GatehouseEvent) -> None:
        delegation_id = event.data["delegation_id"]
        logger.info("Handling delegation.revoked: %s", delegation_id)
        await self._delegation_cache.update_status(delegation_id, "revoked")

    async def _handle_delegation_expired(self, event: GatehouseEvent) -> None:
        delegation_id = event.data["delegation_id"]
        logger.info("Handling delegation.expired: %s", delegation_id)
        await self._delegation_cache.update_status(delegation_id, "expired")

    async def _handle_delegation_exhausted(self, event: GatehouseEvent) -> None:
        delegation_id = event.data["delegation_id"]
        logger.info("Handling delegation.exhausted: %s", delegation_id)
        await self._delegation_cache.update_status(delegation_id, "exhausted")

    async def _handle_agent_deactivated(self, event: GatehouseEvent) -> None:
        agent_id = event.data["agent_id"]
        logger.info("Handling agent.deactivated: %s", agent_id)
        await self._delegation_cache.revoke_all_for_agent(agent_id)

File: src/events/handler.ts
/**
 * Event handler — processes Citadel and Sphinx events to keep
 * local caches in sync.
 */

import { MembershipCache } from '../membership/cache';
import { DelegationCache } from '../delegation/cache';
import { RoleRepository } from '../roles/repository';
import { RoleAssignmentManager } from '../roles/assignment';
import { PermissionResolver } from '../roles/resolver';
import { GatehouseEvent, MetricsCollector, EntityType } from '../types';
import { ResolvedConfig } from '../config';
import * as EventTypes from './types';
import { createLogger } from '../logger';

const logger = createLogger('event-handler');

export class EventHandlerRegistry {
  constructor(
    private membershipCache: MembershipCache,
    private delegationCache: DelegationCache,
    private roleRepo: RoleRepository,
    private roleAssignments: RoleAssignmentManager,
    private permissionResolver: PermissionResolver,
    private config: ResolvedConfig,
    private metrics?: MetricsCollector,
  ) {}

  /**
   * Process an incoming event. Idempotent by design.
   */
  async handle(event: GatehouseEvent): Promise<void> {
    const start = Date.now();

    try {
      switch (event.type) {
        // ─── Organization Events ────────────────────────────────
        case EventTypes.ORG_CREATED:
          await this.handleOrgCreated(event);
          break;
        case EventTypes.ORG_DELETED:
          await this.handleOrgDeleted(event);
          break;
        case EventTypes.ORG_SUSPENDED:
          await this.handleOrgSuspended(event);
          break;
        case EventTypes.ORG_REACTIVATED:
          await this.handleOrgReactivated(event);
          break;

        // ─── Membership Events ──────────────────────────────────
        case EventTypes.MEMBERSHIP_CREATED:
          await this.handleMembershipCreated(event);
          break;
        case EventTypes.MEMBERSHIP_UPDATED:
          await this.handleMembershipUpdated(event);
          break;
        case EventTypes.MEMBERSHIP_SUSPENDED:
          await this.handleMembershipSuspended(event);
          break;
        case EventTypes.MEMBERSHIP_REACTIVATED:
          await this.handleMembershipReactivated(event);
          break;
        case EventTypes.MEMBERSHIP_REMOVED:
          await this.handleMembershipRemoved(event);
          break;

        // ─── Group Events ───────────────────────────────────────
        case EventTypes.GROUP_MEMBER_ADDED:
          await this.handleGroupMemberAdded(event);
          break;
        case EventTypes.GROUP_MEMBER_REMOVED:
          await this.handleGroupMemberRemoved(event);
          break;
        case EventTypes.GROUP_DELETED:
          await this.handleGroupDeleted(event);
          break;

        // ─── Delegation Events ──────────────────────────────────
        case EventTypes.DELEGATION_CREATED:
          await this.handleDelegationCreated(event);
          break;
        case EventTypes.DELEGATION_REVOKED:
          await this.handleDelegationRevoked(event);
          break;
        case EventTypes.DELEGATION_EXPIRED:
          await this.handleDelegationExpired(event);
          break;
        case EventTypes.DELEGATION_EXHAUSTED:
          await this.handleDelegationExhausted(event);
          break;
        case EventTypes.AGENT_DEACTIVATED:
          await this.handleAgentDeactivated(event);
          break;

        default:
          logger.debug({ type: event.type }, 'Unknown event type, ignoring');
      }

      this.metrics?.increment('gatedhouse_event_processed_total', {
        type: event.type,
      });
    } catch (err) {
      logger.error({ err, event: event.type }, 'Event handling failed');
      throw err;
    } finally {
      const duration = Date.now() - start;
      this.metrics?.observe('gatedhouse_event_processing_lag_ms', duration, {
        type: event.type,
      });
    }
  }

  // ─── Organization Handlers ──────────────────────────────────────

  private async handleOrgCreated(event: GatehouseEvent): Promise<void> {
    const orgId = event.data.org_id as string;
    logger.info({ orgId }, 'Handling org.created');
    await this.roleRepo.seedBaseRoles(orgId, this.config.baseRoles);
  }

  private async handleOrgDeleted(event: GatehouseEvent): Promise<void> {
    const orgId = event.data.org_id as string;
    logger.info({ orgId }, 'Handling org.deleted — purging all data');

    await this.membershipCache.removeAllForOrg(orgId);
    await this.roleAssignments.deleteAllForOrg(orgId);
    await this.roleAssignments.deleteAllGroupRolesForOrg(orgId);
    await this.roleRepo.deleteAllForOrg(orgId);
    await this.delegationCache.removeAllForOrg(orgId);
  }

  private async handleOrgSuspended(event: GatehouseEvent): Promise<void> {
    const orgId = event.data.org_id as string;
    logger.info({ orgId }, 'Handling org.suspended');
    await this.membershipCache.suspendAllForOrg(orgId);
  }

  private async handleOrgReactivated(event: GatehouseEvent): Promise<void> {
    const orgId = event.data.org_id as string;
    logger.info({ orgId }, 'Handling org.reactivated');
    await this.membershipCache.reactivateAllForOrg(orgId);
  }

  // ─── Membership Handlers ───────────────────────────────────────

  private async handleMembershipCreated(event: GatehouseEvent): Promise<void> {
    const data = event.data;
    const membershipId = data.membership_id as string;
    const orgId = data.org_id as string;

    logger.info({ membershipId, orgId }, 'Handling membership.created');

    await this.membershipCache.upsert({
      membershipId,
      orgId,
      entityType: data.entity_type as EntityType,
      entityId: data.entity_id as string,
      isOwner: (data.is_owner as boolean) ?? false,
      status: (data.status as string) ?? 'active',
      groups: (data.groups as string[]) ?? [],
    });

    // Assign default role
    if (this.config.defaultRole) {
      await this.roleAssignments.assign(
        membershipId,
        this.config.defaultRole,
        orgId,
      );
    }

    // If owner, also assign owner role
    if (data.is_owner) {
      await this.roleAssignments.assign(membershipId, 'owner', orgId);
    }

    // Build initial resolved permissions
    await this.permissionResolver.rebuildForMembership(
      membershipId,
      orgId,
      (data.groups as string[]) ?? [],
    );
  }

  private async handleMembershipUpdated(event: GatehouseEvent): Promise<void> {
    const data = event.data;
    const membershipId = data.membership_id as string;
    const orgId = data.org_id as string;

    logger.info({ membershipId }, 'Handling membership.updated');

    await this.membershipCache.upsert({
      membershipId,
      orgId,
      entityType: data.entity_type as EntityType,
      entityId: data.entity_id as string,
      isOwner: (data.is_owner as boolean) ?? false,
      status: (data.status as string) ?? 'active',
      groups: (data.groups as string[]) ?? [],
    });

    // Rebuild permissions in case owner status changed
    await this.permissionResolver.rebuildForMembership(
      membershipId,
      orgId,
      (data.groups as string[]) ?? [],
    );
  }

  private async handleMembershipSuspended(event: GatehouseEvent): Promise<void> {
    const membershipId = event.data.membership_id as string;
    logger.info({ membershipId }, 'Handling membership.suspended');
    await this.membershipCache.updateStatus(membershipId, 'suspended');
  }

  private async handleMembershipReactivated(event: GatehouseEvent): Promise<void> {
    const membershipId = event.data.membership_id as string;
    logger.info({ membershipId }, 'Handling membership.reactivated');
    await this.membershipCache.updateStatus(membershipId, 'active');
  }

  private async handleMembershipRemoved(event: GatehouseEvent): Promise<void> {
    const membershipId = event.data.membership_id as string;
    logger.info({ membershipId }, 'Handling membership.removed');

    await this.membershipCache.remove(membershipId);
    await this.roleAssignments.deleteAllForMembership(membershipId);
    await this.permissionResolver.clearForMembership(membershipId);
  }

  // ─── Group Handlers ────────────────────────────────────────────

  private async handleGroupMemberAdded(event: GatehouseEvent): Promise<void> {
    const membershipId = event.data.membership_id as string;
    const groupId = event.data.group_id as string;
    const orgId = event.data.org_id as string;

    logger.info({ membershipId, groupId }, 'Handling group.member.added');

    await this.membershipCache.addGroup(membershipId, groupId);

    // Rebuild permissions to include group roles
    const cached = await this.membershipCache.findById(membershipId);
    if (cached) {
      await this.permissionResolver.rebuildForMembership(
        membershipId,
        orgId,
        cached.groups,
      );
    }
  }

  private async handleGroupMemberRemoved(event: GatehouseEvent): Promise<void> {
    const membershipId = event.data.membership_id as string;
    const groupId = event.data.group_id as string;
    const orgId = event.data.org_id as string;

    logger.info({ membershipId, groupId }, 'Handling group.member.removed');

    await this.membershipCache.removeGroup(membershipId, groupId);

    // Rebuild permissions without group roles
    const cached = await this.membershipCache.findById(membershipId);
    if (cached) {
      await this.permissionResolver.rebuildForMembership(
        membershipId,
        orgId,
        cached.groups,
      );
    }
  }

  private async handleGroupDeleted(event: GatehouseEvent): Promise<void> {
    const groupId = event.data.group_id as string;
    const orgId = event.data.org_id as string;

    logger.info({ groupId }, 'Handling group.deleted');

    // Remove group from all membership caches
    await this.membershipCache.removeGroupFromAll(groupId);

    // Remove group role assignments
    await this.roleAssignments.deleteAllForGroup(groupId);

    // Rebuild permissions for all affected memberships
    const memberships = await this.membershipCache.listByOrg(orgId);
    for (const m of memberships) {
      await this.permissionResolver.rebuildForMembership(
        m.membershipId,
        orgId,
        m.groups,
      );
    }
  }

  // ─── Delegation Handlers ───────────────────────────────────────

  private async handleDelegationCreated(event: GatehouseEvent): Promise<void> {
    const data = event.data;
    logger.info(
      { delegationId: data.delegation_id },
      'Handling delegation.created',
    );

    await this.delegationCache.upsert({
      delegationId: data.delegation_id as string,
      agentId: data.agent_id as string,
      delegatorId: data.delegator_id as string,
      delegatorMembershipId: data.delegator_membership_id as string,
      orgId: data.org_id as string,
      scopes: (data.scopes as string[]) ?? [],
      constraints: (data.constraints as Record<string, unknown>) ?? {},
      maxUses: (data.max_uses as number) ?? null,
      useCount: 0,
      status: 'active',
      expiresAt: new Date(data.expires_at as string),
    });
  }

  private async handleDelegationRevoked(event: GatehouseEvent): Promise<void> {
    const delegationId = event.data.delegation_id as string;
    logger.info({ delegationId }, 'Handling delegation.revoked');
    await this.delegationCache.updateStatus(delegationId, 'revoked');
  }

  private async handleDelegationExpired(event: GatehouseEvent): Promise<void> {
    const delegationId = event.data.delegation_id as string;
    logger.info({ delegationId }, 'Handling delegation.expired');
    await this.delegationCache.updateStatus(delegationId, 'expired');
  }

  private async handleDelegationExhausted(event: GatehouseEvent): Promise<void> {
    const delegationId = event.data.delegation_id as string;
    logger.info({ delegationId }, 'Handling delegation.exhausted');
    await this.delegationCache.updateStatus(delegationId, 'exhausted');
  }

  private async handleAgentDeactivated(event: GatehouseEvent): Promise<void> {
    const agentId = event.data.agent_id as string;
    logger.info({ agentId }, 'Handling agent.deactivated');
    await this.delegationCache.revokeAllForAgent(agentId);
  }
}
